server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -196,13 +200,26 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final
     final LinkedHashMap<SegmentDescriptor, List<QueryRunner<T>>> allRunners = new LinkedHashMap<>();
 
     try {
-      for (final SegmentDescriptor newDescriptor : specs) {
-        final SegmentDescriptor descriptor = newIdToBasePendingSegment.getOrDefault(newDescriptor, newDescriptor);
-        final PartitionChunk<Sink> chunk = sinkTimeline.findChunk(
+      for (final SegmentDescriptor descriptor : specs) {
+        final PartitionChunk<SegmentDescriptorOvershadowable> descriptorChunk = upgradedSegmentsTimeline.findChunk(
             descriptor.getInterval(),
             descriptor.getVersion(),
             descriptor.getPartitionNumber()
         );
+        final SegmentDescriptor baseDescriptor;
+        if (descriptorChunk != null) {
+          baseDescriptor = newIdToBasePendingSegment.getOrDefault(
+              descriptorChunk.getObject().segmentDescriptor,
+              descriptor
+          );
+        } else {
+          baseDescriptor = descriptor;
+        }
+        final PartitionChunk<Sink> chunk = sinkTimeline.findChunk(
+            baseDescriptor.getInterval(),
```

Comment:
This is in some cases going to search for a wider interval than the original descriptor, because descriptors can have narrower intervals than the segments they reference. This was missing from the javadoc of `SegmentDescriptor`, so I added it in: #17274

At any rate, I think it's possible for the original `descriptor` to have a narrower interval than `baseDescriptor` due to this effect. Is that going to be a problem here?

Reply:
The baseDescriptor would be at most as wide as the actual segment's interval.
I believe this shouldn't be a problem as the baseDescriptor is only being used to determine the sink.